src/AssocArray.php

```diff
@@ -1,45 +1,55 @@
-<?php
+<?php declare(strict_types=1);
 
 namespace GW\Value;
 
 use RuntimeException;
 use ArrayIterator;
 use BadMethodCallException;
+use function array_keys;
+use function array_map;
 use function in_array;
 use function count;
 
+/**
+ * @template TKey
+ * @template TValue
+ * @implements AssocValue<TKey, TValue>
+ */
 final class AssocArray implements AssocValue
 {
+    /** @phpstan-var array<TKey, TValue> */
     private array $items;
 
+    /**
+     * @phpstan-param array<TKey, TValue> $items
+     */
     public function __construct(array $items)
     {
         $this->items = $items;
     }
 
     /**
-     * @param callable $transformer function(mixed $value[, string $key]): mixed
+     * @template TNewValue
+     * @phpstan-param callable(TValue $value, TKey $key=):TNewValue $transformer
+     * @phpstan-return AssocArray<TKey, TNewValue>
      */
     public function map(callable $transformer): AssocArray
     {
-        $result = [];
-        foreach ($this->items as $key => $value) {
-            $result[$key] = $transformer($value, $key);
-        }
-
-        return new self($result);
+        return new self(array_map($transformer, $this->items, array_keys($this->items)));
     }
 
     /**
-     * @return StringsArray
+     * @phpstan-return ArrayValue<TKey>
      */
-    public function keys(): StringsArray
+    public function keys(): ArrayValue
     {
-        return Wrap::stringsArray(array_keys($this->items));
+        return Wrap::array(array_keys($this->items));
     }
 
     /**
-     * @param callable $transformer function(string $key[, mixed $value]): string
+     * @template TNewKey
+     * @param callable(TKey $key, TValue $value=): TNewKey $transformer
```

Comment:
Here is also redundant-looking `=`. Or is it some convention to mark optional parameter?

Reply:
it's convention to mark parameter as optional :) 

unfortunately, there is no information about it in phpstan docs - only in psalm:
https://psalm.dev/docs/annotating_code/type_syntax/callable_types/